In [7]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [2]:
from google.colab import files # Only if using Google Colab
uploaded = files.upload() # This will prompt you to select the file

Saving EDA_output.csv to EDA_output.csv


In [3]:
data = pd.read_csv('EDA_output.csv')
print(data.head())

  diagnosis  perimeter_mean  concave_points_mean  fractal_dimension_mean  \
0         M          122.80              0.14710                 0.07871   
1         M          132.90              0.07017                 0.05667   
2         M          130.00              0.12790                 0.05999   
3         M           77.58              0.10520                 0.09744   
4         M          135.10              0.10430                 0.05883   

   texture_se  area_se  smoothness_se  concavity_se  symmetry_se  \
0      0.9053   153.40       0.006399       0.05373      0.03003   
1      0.7339    74.08       0.005225       0.01860      0.01389   
2      0.7869    94.03       0.006150       0.03832      0.02250   
3      1.1560    27.23       0.009110       0.05661      0.05963   
4      0.7813    94.44       0.011490       0.05688      0.01756   

   fractal_dimension_se  texture_worst  area_worst  smoothness_worst  \
0              0.006193          17.33      2019.0            

In [4]:
categorical_columns = data.select_dtypes(include=['object']).columns
for col in categorical_columns:
    le = LabelEncoder()
    data[col] = le.fit_transform(data[col])

# Separate features and target variable
X = data.drop('diagnosis', axis=1)
y = data['diagnosis']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [5]:
#1.logistic regression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred_logistic = logistic_model.predict(X_test)
print("Logistic Regression:")
print(classification_report(y_test, y_pred_logistic))

Logistic Regression:
              precision    recall  f1-score   support

           0       0.97      0.98      0.98       108
           1       0.97      0.95      0.96        63

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [6]:
#2.SVM
svm_model = SVC()
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)
print("SVM:")
print(classification_report(y_test, y_pred_svm))

SVM:
              precision    recall  f1-score   support

           0       0.93      0.99      0.96       108
           1       0.98      0.87      0.92        63

    accuracy                           0.95       171
   macro avg       0.96      0.93      0.94       171
weighted avg       0.95      0.95      0.95       171



In [7]:
#3. KNN
knn_model = KNeighborsClassifier(n_neighbors=5)  # k=5 can be tuned
knn_model.fit(X_train, y_train)
y_pred_knn = knn_model.predict(X_test)
print("KNN:")
print(classification_report(y_test, y_pred_knn))

KNN:
              precision    recall  f1-score   support

           0       0.94      0.94      0.94       108
           1       0.90      0.89      0.90        63

    accuracy                           0.92       171
   macro avg       0.92      0.92      0.92       171
weighted avg       0.92      0.92      0.92       171



In [8]:
# 4. Neural Network
nn_model = Sequential()
nn_model.add(Dense(128, input_dim=X_train.shape[1], activation='relu'))
nn_model.add(Dense(64, activation='relu'))
nn_model.add(Dense(1, activation='sigmoid'))  # Binary classification

nn_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the Neural Network
nn_model.fit(X_train, y_train, epochs=50, batch_size=10, verbose=1)

# Predict with the Neural Network
y_pred_nn = (nn_model.predict(X_test) > 0.5).astype("int32")
print("Neural Network:")
print(classification_report(y_test, y_pred_nn))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6239 - loss: 3.3502
Epoch 2/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8065 - loss: 0.4294
Epoch 3/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8255 - loss: 0.4656
Epoch 4/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8371 - loss: 0.4953
Epoch 5/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8447 - loss: 0.8283
Epoch 6/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8615 - loss: 0.3486
Epoch 7/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8788 - loss: 0.3838
Epoch 8/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8503 - loss: 0.4078
Epoch 9/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7864 - loss: 1.0282
Epoch 10/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8180 - loss: 0.5187
Epoch 11/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8687 - loss: 0.6187
Epoch 12/50
40/40 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8344 - lo

In [9]:
#Accuracy of each model
print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_logistic))
print("SVM Accuracy:", accuracy_score(y_test, y_pred_svm))
print("KNN Accuracy:", accuracy_score(y_test, y_pred_knn))
print("Neural Network Accuracy:", accuracy_score(y_test, y_pred_nn))


Logistic Regression Accuracy: 0.9707602339181286
SVM Accuracy: 0.9473684210526315
KNN Accuracy: 0.9239766081871345
Neural Network Accuracy: 0.9766081871345029


Summary of Performance Across Models:
From the provided precision, recall, and F1-scores for each model, we can draw the following conclusions:

1. Logistic Regression:
Accuracy: 97%

Precision:

Class 0 (Benign): 97%
Class 1 (Malignant): 97%
Recall:

Class 0 (Benign): 98%
Class 1 (Malignant): 95%
F1-Score:

Class 0 (Benign): 98%
Class 1 (Malignant): 96%
Interpretation: Logistic Regression performs very well overall, with a balanced performance across both classes. The recall for malignant cases is 95%, meaning it detects 95% of the actual malignant cases, which is important in this context.

2. SVM:
Accuracy: 95%

Precision:

Class 0 (Benign): 93%
Class 1 (Malignant): 98%
Recall:

Class 0 (Benign): 99%
Class 1 (Malignant): 87%
F1-Score:

Class 0 (Benign): 96%
Class 1 (Malignant): 92%
Interpretation: SVM has a high precision for detecting malignant cases (98%) but a lower recall (87%), meaning it misses some malignant cases. This may be critical in medical diagnosis, where missing malignant cases could have serious consequences.

3. KNN:
Accuracy: 92%

Precision:

Class 0 (Benign): 94%
Class 1 (Malignant): 90%
Recall:

Class 0 (Benign): 94%
Class 1 (Malignant): 89%
F1-Score:

Class 0 (Benign): 94%
Class 1 (Malignant): 90%
Interpretation: KNN performs reasonably well, but its recall and precision for malignant cases are lower compared to Logistic Regression and Neural Networks. This suggests that KNN may not be the best choice for this task.

4. Neural Network:
Accuracy: 98%

Precision:

Class 0 (Benign): 98%
Class 1 (Malignant): 97%
Recall:

Class 0 (Benign): 98%
Class 1 (Malignant): 97%
F1-Score:

Class 0 (Benign): 98%
Class 1 (Malignant): 97%
Interpretation: The Neural Network achieves the highest accuracy (98%) with balanced precision, recall, and F1-scores for both classes. This suggests it performs well at both detecting and correctly identifying benign and malignant cases.

In [18]:
from sklearn.model_selection import KFold
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy as np

# Define the neural network model
def create_nn_model():
    model = Sequential()
    model.add(Dense(128, input_dim=X_train.shape[1], activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))  # Binary classification
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

for train_index, val_index in kf.split(X_train):
    # Use .iloc for integer indexing on pandas DataFrames
    X_train_fold, X_val_fold = X_train.iloc[train_index], X_train.iloc[val_index]
    y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[val_index]

    # Build and train the model
    model = create_nn_model()
    model.fit(X_train_fold, y_train_fold, epochs=50, batch_size=10, verbose=0)

    # Evaluate the model
    scores = model.evaluate(X_val_fold, y_val_fold, verbose=0)
    print(f"Fold {fold_no} - Accuracy: {scores[1]}")
    accuracy_scores.append(scores[1])
    fold_no += 1

# Print mean and standard deviation of accuracy
print(f"Mean Accuracy: {np.mean(accuracy_scores)}, Standard Deviation: {np.std(accuracy_scores)}")


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Fold 1 - Accuracy: 0.7875000238418579


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Fold 2 - Accuracy: 0.9125000238418579


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Fold 3 - Accuracy: 0.887499988079071


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Fold 4 - Accuracy: 0.8734177350997925


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Fold 5 - Accuracy: 0.9240506291389465
Mean Accuracy: 0.8769936800003052, Standard Deviation: 0.04818161744735243


In [14]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

# Step 1: Load your dataset
# Make sure to update this with the correct file path or dataset
# Assuming 'data' is a pandas DataFrame and 'target' is the column with labels
data = pd.read_csv('EDA_output.csv')
print(data.head())

# Step 2: Separate features (X) and target (y)
X = data.drop('diagnosis', axis=1)  # Replace 'target' with the actual column name
y = data['diagnosis']  # Replace 'target' with the actual column name

# Step 3: Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 4: Perform hyperparameter tuning (Logistic Regression example)
logistic_params = {
    'C': [0.1, 1, 10],
    'solver': ['liblinear', 'lbfgs']
}
logistic_grid = GridSearchCV(LogisticRegression(), logistic_params, cv=5, scoring='accuracy')
logistic_grid.fit(X_train, y_train)
best_logistic = logistic_grid.best_estimator_  # Best Logistic Regression model

# Perform hyperparameter tuning for SVM
svm_params = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale', 'auto']
}
svm_grid = GridSearchCV(SVC(), svm_params, cv=5, scoring='accuracy')
svm_grid.fit(X_train, y_train)
best_svm = svm_grid.best_estimator_

# Perform hyperparameter tuning for KNN
knn_params = {
    'n_neighbors': [3, 5, 7],
    'weights': ['uniform', 'distance'],
    'p': [1, 2]
}
knn_grid = GridSearchCV(KNeighborsClassifier(), knn_params, cv=5, scoring='accuracy')
knn_grid.fit(X_train, y_train)
best_knn = knn_grid.best_estimator_

# Now you can perform cross-validation using the best models

from sklearn.model_selection import cross_val_score

# Dictionary of best models after hyperparameter tuning
models = {
    'Logistic Regression': best_logistic,
    'SVM': best_svm,
    'KNN': best_knn
}

# Perform cross-validation and evaluate each model
for name, model in models.items():
    print(f"Performing cross-validation for {name}...")

    # Perform 5-fold cross-validation
    scores = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')

    # Output cross-validation results
    print(f"Cross-validation scores for {name}: {scores}")
    print(f"Mean accuracy: {scores.mean():.4f}")
    print(f"Standard deviation: {scores.std():.4f}")
    print("-" * 40)

  diagnosis  perimeter_mean  concave_points_mean  fractal_dimension_mean  \
0         M          122.80              0.14710                 0.07871   
1         M          132.90              0.07017                 0.05667   
2         M          130.00              0.12790                 0.05999   
3         M           77.58              0.10520                 0.09744   
4         M          135.10              0.10430                 0.05883   

   texture_se  area_se  smoothness_se  concavity_se  symmetry_se  \
0      0.9053   153.40       0.006399       0.05373      0.03003   
1      0.7339    74.08       0.005225       0.01860      0.01389   
2      0.7869    94.03       0.006150       0.03832      0.02250   
3      1.1560    27.23       0.009110       0.05661      0.05963   
4      0.7813    94.44       0.011490       0.05688      0.01756   

   fractal_dimension_se  texture_worst  area_worst  smoothness_worst  \
0              0.006193          17.33      2019.0            

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Performing cross-validation for Logistic Regression...
Cross-validation scores for Logistic Regression: [0.97802198 0.94505495 0.97802198 0.96703297 0.92307692]
Mean accuracy: 0.9582
Standard deviation: 0.0213
----------------------------------------
Performing cross-validation for SVM...


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Cross-validation scores for SVM: [0.96703297 0.97802198 0.97802198 0.97802198 0.92307692]
Mean accuracy: 0.9648
Standard deviation: 0.0213
----------------------------------------
Performing cross-validation for KNN...
Cross-validation scores for KNN: [0.94505495 0.87912088 0.96703297 0.89010989 0.91208791]
Mean accuracy: 0.9187
Standard deviation: 0.0330
----------------------------------------


Cross-Validation Results:
SVM:

Mean Accuracy: 96.48%
Standard Deviation: 2.13%
Interpretation: SVM performs well, showing high accuracy and low variation across folds. However, in one fold, the accuracy dropped to 92.31%, which is lower than the others. This may indicate some instability in certain parts of the data.
KNN:

Mean Accuracy: 91.87%
Standard Deviation: 3.30%
Interpretation: KNN has lower accuracy compared to SVM, with more variation across folds. The standard deviation is relatively higher, suggesting that KNN's performance is more sensitive to how the data is split.
Logistic Regression:

Mean Accuracy: 95.82%
Standard Deviation: 2.13%
Interpretation: Logistic Regression performs quite well, similar to SVM. The standard deviation indicates stable performance across different folds.
Neural Network:

Mean Accuracy: 87.70%
Standard Deviation: 4.82%
Interpretation: The neural network has the lowest mean accuracy and the highest standard deviation, indicating that its performance is less stable across folds compared to other models. This may suggest overfitting or underfitting, depending on the network's architecture or hyperparameters.

In [8]:
data = pd.read_csv('/Users/lakshikajain/Documents/GitHub/DataScienceCapstone_3/Untitled/EDA_output.csv')
print(data.head()) 

  diagnosis  perimeter_mean  concave_points_mean  fractal_dimension_mean  \
0         M          122.80              0.14710                 0.07871   
1         M          132.90              0.07017                 0.05667   
2         M          130.00              0.12790                 0.05999   
3         M           77.58              0.10520                 0.09744   
4         M          135.10              0.10430                 0.05883   

   texture_se  area_se  smoothness_se  concavity_se  symmetry_se  \
0      0.9053   153.40       0.006399       0.05373      0.03003   
1      0.7339    74.08       0.005225       0.01860      0.01389   
2      0.7869    94.03       0.006150       0.03832      0.02250   
3      1.1560    27.23       0.009110       0.05661      0.05963   
4      0.7813    94.44       0.011490       0.05688      0.01756   

   fractal_dimension_se  texture_worst  area_worst  smoothness_worst  \
0              0.006193          17.33      2019.0            

In [11]:
from pytorch_tabnet.tab_model import TabNetClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Assuming your dataset is loaded into a pandas DataFrame
# Replace 'target' with the actual name of your target column
X = data.drop('diagnosis', axis=1)
y = data['diagnosis']

# Encode the target if it's categorical
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

# Split the dataset into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X.values, y, test_size=0.2, random_state=42)

# Define the TabNet model for binary classification
model = TabNetClassifier()

# Train the model with transfer learning
model.fit(
    X_train, y_train,
    eval_set=[(X_val, y_val)],
    max_epochs=100,
    patience=10,
    batch_size=1024,
    virtual_batch_size=128,
    from_unsupervised=None  # This is where you can specify a pre-trained model for transfer learning
)

# Predict and evaluate
y_pred = model.predict(X_val)


/opt/anaconda3/lib/python3.11/site-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cpu
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 1  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 2  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 3  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 4  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 5  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 6  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 7  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 8  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 9  | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s
epoch 10 | loss: 0.0     | val_0_auc: 0.71782 |  0:00:00s

Early stopping occurred at epoch 10 with best_epoch = 0 and best_val_0_auc = 0.71782


/opt/anaconda3/lib/python3.11/site-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


In [15]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
# Build a simple neural network model
base_model = Sequential([
    Dense(64, input_dim=X_train.shape[1], activation='relu'),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile the model
base_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
base_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))


Epoch 1/10


/opt/anaconda3/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5144 - loss: 8.5206 - val_accuracy: 0.8947 - val_loss: 0.2973
Epoch 2/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7576 - loss: 0.7344 - val_accuracy: 0.9240 - val_loss: 0.1878
Epoch 3/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8913 - loss: 0.2851 - val_accuracy: 0.8480 - val_loss: 0.3663
Epoch 4/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8344 - loss: 0.3777 - val_accuracy: 0.9357 - val_loss: 0.1615
Epoch 5/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9399 - loss: 0.1679 - val_accuracy: 0.9649 - val_loss: 0.1068
Epoch 6/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8939 - loss: 0.2674 - val_accuracy: 0.9298 - val_loss: 0.1462
Epoch 7/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9509 - loss: 0.1542 - val_accuracy: 0.9474 - val_loss: 0.1278
Epoch 8/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9036 - loss: 0.2521 - val_accuracy: 0.9415 - val_loss: 0.1444
Epo

In [17]:
# Freeze the layers of the base model
for layer in base_model.layers[:-1]:
    layer.trainable = False

# Fine-tune the model by adding a new layer for further learning
fine_tune_model = Sequential(base_model.layers)
fine_tune_model.add(Dense(1, activation='sigmoid'))  # Adding a new output layer

# Compile the fine-tuned model
fine_tune_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Fine-tune the model
fine_tune_model.fit(X_train, y_train, epochs=5, batch_size=32, validation_data=(X_test, y_test))


Epoch 1/5
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6286 - loss: 0.8513 - val_accuracy: 0.6316 - val_loss: 0.6920
Epoch 2/5
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6400 - loss: 0.6914 - val_accuracy: 0.6316 - val_loss: 0.6903
Epoch 3/5
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6336 - loss: 0.6900 - val_accuracy: 0.6316 - val_loss: 0.6890
Epoch 4/5
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6601 - loss: 0.6877 - val_accuracy: 0.6316 - val_loss: 0.6877
Epoch 5/5
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6324 - loss: 0.6874 - val_accuracy: 0.6316 - val_loss: 0.6865


In [19]:
# Evaluate the fine-tuned model
loss, accuracy = fine_tune_model.evaluate(X_test, y_test)
print(f"Test Accuracy: {accuracy * 100:.2f}%")


6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 888us/step - accuracy: 0.6357 - loss: 0.6863
Test Accuracy: 63.16%


In [27]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

# Define the model building function for tuning
def create_model(learning_rate=0.001, dropout_rate=0.0):
    model = Sequential()
    model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))  # First hidden layer
    model.add(Dense(32, activation='relu'))  # Second hidden layer
    model.add(Dropout(dropout_rate))  # Dropout for regularization
    model.add(Dense(1, activation='sigmoid'))  # Output layer for binary classification

    # Compile the model
    optimizer = Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])
    
    return model


In [31]:
import tensorflow as tf
from tensorflow import keras
from keras_tuner import RandomSearch
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.models import Sequential

# Define a model building function for Keras Tuner
def build_model(hp):
    model = Sequential()
    model.add(Dense(units=hp.Int('units', min_value=32, max_value=512, step=32), activation='relu', input_dim=X_train.shape[1]))
    model.add(Dense(units=hp.Int('units_2', min_value=32, max_value=512, step=32), activation='relu'))
    model.add(Dropout(rate=hp.Float('dropout_rate', min_value=0.0, max_value=0.5, step=0.1)))
    model.add(Dense(1, activation='sigmoid'))

    # Compile the model
    model.compile(optimizer=Adam(learning_rate=hp.Choice('learning_rate', values=[1e-2, 1e-3, 1e-4])),
                  loss='binary_crossentropy',
                  metrics=['accuracy'])
    return model

# Initialize the tuner
tuner = RandomSearch(
    build_model,
    objective='val_accuracy',
    max_trials=5,  # Number of hyperparameter combinations to try
    executions_per_trial=3,  # Number of models built and trained for each trial
    directory='my_dir',
    project_name='hyperparameter_tuning')

# Search for the best hyperparameters
tuner.search(X_train, y_train, epochs=10, validation_data=(X_test, y_test))

# Get the optimal hyperparameters
best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]
print(f"Best hyperparameters: {best_hps}")

# Build and train the final model
model = tuner.hypermodel.build(best_hps)
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))


Trial 5 Complete [00h 00m 04s]
val_accuracy: 0.9493177533149719

Best val_accuracy So Far: 0.9493177533149719
Total elapsed time: 00h 00m 21s
Best hyperparameters: <keras_tuner.src.engine.hyperparameters.hyperparameters.HyperParameters object at 0x164c28f50>
Epoch 1/10


/opt/anaconda3/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


13/13 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - accuracy: 0.5063 - loss: 101.0625 - val_accuracy: 0.3684 - val_loss: 6.0794
Epoch 2/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5753 - loss: 4.7028 - val_accuracy: 0.9298 - val_loss: 0.1955
Epoch 3/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8294 - loss: 0.4474 - val_accuracy: 0.9064 - val_loss: 0.2285
Epoch 4/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8666 - loss: 0.3744 - val_accuracy: 0.9123 - val_loss: 0.1924
Epoch 5/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8580 - loss: 0.3406 - val_accuracy: 0.9474 - val_loss: 0.1580
Epoch 6/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8854 - loss: 0.2700 - val_accuracy: 0.9415 - val_loss: 0.2079
Epoch 7/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8928 - loss: 0.2944 - val_accuracy: 0.8947 - val_loss: 0.2156
Epoch 8/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8916 - loss: 0.2997 - val_accuracy: 0.9474 - val_loss: 0.1392
E

In [33]:
best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]
print(best_hps.values)


{'units': 224, 'units_2': 512, 'dropout_rate': 0.2, 'learning_rate': 0.01}


In [35]:
# Rebuild the model using the best hyperparameters
final_model = create_model(learning_rate=0.01, dropout_rate=0.2)

# Train the final model
history = final_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))

# Evaluate the final model
loss, accuracy = final_model.evaluate(X_test, y_test)
print(f"Final Model Test Accuracy: {accuracy * 100:.2f}%")


Epoch 1/10


/opt/anaconda3/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


13/13 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - accuracy: 0.4850 - loss: 30.7002 - val_accuracy: 0.7661 - val_loss: 2.6084
Epoch 2/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6106 - loss: 6.5592 - val_accuracy: 0.8538 - val_loss: 0.5617
Epoch 3/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7205 - loss: 1.3941 - val_accuracy: 0.8012 - val_loss: 0.3554
Epoch 4/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8000 - loss: 0.4116 - val_accuracy: 0.9064 - val_loss: 0.2165
Epoch 5/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8446 - loss: 0.3711 - val_accuracy: 0.9415 - val_loss: 0.2243
Epoch 6/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8500 - loss: 0.3768 - val_accuracy: 0.8889 - val_loss: 0.3482
Epoch 7/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8324 - loss: 0.4664 - val_accuracy: 0.9181 - val_loss: 0.4363
Epoch 8/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8736 - loss: 0.4644 - val_accuracy: 0.9240 - val_loss: 0.3056
Ep